### PR TITLE
Feature/consume with high lever consumer

### PR DESCRIPTION
**Pull request by NastasiaSaby**
First version of the Kafk bundle with the high level consumer. You will find more information in the Readme.md file.

You will also find _Integration Tests_ if you want to test the bundle. It will call Kafka if you have it. I would like to keep them but I do not know how because there is no fixture for the moment and I do not know how to do that for the moment.

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        $rootNode
+            ->children()
+                ->booleanNode('event_dispatcher')->defaultTrue()->end()
+                ->arrayNode('consumers')
+                    ->useAttributeAsKey('key')
+                    ->prototype('array')
+                        ->children()
+                            ->arrayNode('conf')
+                                ->prototype('scalar')->end()
+                                    ->defaultValue(array())
```
Comment (twdmnd) [inline DependencyInjection/Configuration.php]: Indentation

Comment (gmillet) [inline DependencyInjection/Configuration.php]: Indentation

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                                    ->normalizeKeys(false)
```
Comment (gmillet) [inline DependencyInjection/Configuration.php]: Indentation

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                            ->end()
+                            ->arrayNode('topicConf')
+                                ->prototype('scalar')->end()
+                                    ->defaultValue(array())
```
Comment (gmillet) [inline DependencyInjection/Configuration.php]: Indentation

File: DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                                    ->normalizeKeys(false)
```
Comment (gmillet) [inline DependencyInjection/Configuration.php]: Indentation

File: DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                        ->end()
+                    ->end()
+                ->end()
+                ->arrayNode('producers')
+                    ->useAttributeAsKey('key')
+                    ->prototype('array')
+                        ->children()
+                            ->arrayNode('conf')
+                                ->prototype('scalar')->end()
+                                    ->defaultValue(array())
```
Comment (gmillet) [inline DependencyInjection/Configuration.php]: Indentation

File: DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                                    ->normalizeKeys(false)
```
Comment (gmillet) [inline DependencyInjection/Configuration.php]: Indentation

File: DependencyInjection/M6WebKafkaExtension.php [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+    /**
+     * @param array      $entity
+     * @param Definition $entityDefinition
+     */
+    protected function setProducerManager(array $entity, Definition $entityDefinition)
+    {
+        $brokers = implode(',', $entity['brokers']);
+
+        $entityDefinition
```
Comment (twdmnd) [inline DependencyInjection/M6WebKafkaExtension.php]: addMethodCall returns entityDefinition, so you can do:
```
$entityDefinition
    ->addMethodCall('addBrokers', [$brokers])
    ->addMethodCall('setLogLevel', [$entity['log_level']]);
```

File: DependencyInjection/M6WebKafkaExtension.php [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+    /**
+     * @param array $confToSet
+     * @return \RdKafka\TopicConf
+     */
+    protected function getReadyTopicConf(array $confToSet = []): \RdKafka\TopicConf
+    {
+        $topicConf = $this->getTopicConf();
+
+        array_walk($confToSet, function ($confValue, $confIndex) use ($topicConf) {
```
Comment (gmillet) [inline DependencyInjection/M6WebKafkaExtension.php]: `array_walk(array_flip($confToSet), $topicConf->set);`

File: DependencyInjection/M6WebKafkaExtension.php [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @param array            $config
+     */
+    protected function loadProducers(ContainerBuilder $container, array $config)
+    {
+        foreach ($config['producers'] as $key => $producer) {
+            $producerDefinition = $this->getDefinition('M6Web\Bundle\KafkaBundle\RdKafkaProducerManager');
```
Comment (fabdsp) [inline DependencyInjection/M6WebKafkaExtension.php]: Can we use a parameter ? to overload this class ?

File: DependencyInjection/M6WebKafkaExtension.php [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @param array $config
+     */
+    protected function loadConsumers(ContainerBuilder $container, array $config)
+    {
+        foreach ($config['consumers'] as $key => $consumer) {
+            $consumerDefinition = $this->getDefinition('M6Web\Bundle\KafkaBundle\RdKafkaConsumerManager');
```
Comment (fabdsp) [inline DependencyInjection/M6WebKafkaExtension.php]: Can we use a parameter ? to overload this class ?

File: DependencyInjection/M6WebKafkaExtension.php [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+     * @param \RdKafka\Conf $rdKafkaConf
+     * @param array         $confData
+     *
+     * @return \RdKafka\Conf
+     */
+    protected function getReadyRdKafkaConf(array $confToSet): \RdKafka\Conf
+    {
+        $rdKafkaConf = $this->getRdKafkaConf();
+
+        array_walk($confToSet, function ($confValue, $confIndex) use ($rdKafkaConf) {
```
Comment (gmillet) [inline DependencyInjection/M6WebKafkaExtension.php]: `array_walk(array_flip($confToSet), $rdKafkaConf->set)`

File: DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                            ->end()
+                            ->arrayNode('topicConf')
+                                ->prototype('scalar')->end()
+                                    ->defaultValue(array())
+                                    ->normalizeKeys(false)
+                            ->end()
+                            ->scalarNode('timeout_consuming_queue')->defaultValue(1000)->end()
+                            ->arrayNode('topics')
```
Comment (lnahiro) [inline DependencyInjection/Configuration.php:39]: pas possible de mettre la conf des topics ici ? comme pour le producer

File: DependencyInjection/M6WebKafkaExtension.php [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        return $rdKafkaConf;
+    }
+
+    /**
+     * @return Callable
+     */
+    protected function handlePartitionsAssignment(): Callable
+    {
+        return function (\RdKafka\KafkaConsumer $kafka, $error, array $partitions = null) {
+            if ($error === RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS) {
```
Comment (gmillet) [inline DependencyInjection/M6WebKafkaExtension.php]: switch ?

File: IntegrationTests/Units/RdKafkaConsumerManager.php [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        $topicConf = new \RdKafka\TopicConf();
+        $topicConf->set('auto.offset.reset', 'smallest');
+        $conf = new \RdKafka\Conf();
+        $conf->set('group.id', 'myConsumerGroup');
+
+        $conf->set('metadata.broker.list', '127.0.0.1');
+        $conf->setDefaultTopicConf($topicConf);
+        $conf->set('enable.auto.commit', '0');
+
```
Comment (twdmnd) [inline IntegrationTests/Units/RdKafkaConsumerManager.php]: extra line

File: DependencyInjection/M6WebKafkaExtension.php (hunk 10 above)
Comment (lnahiro) [inline DependencyInjection/M6WebKafkaExtension.php]: ConsumerManager ?

File: DependencyInjection/M6WebKafkaExtension.php [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+            $this->setEventDispatcher($config, $consumerDefinition);
+
+            $topicConf = $this->getReadyTopicConf($consumer['topicConf']);
+            $conf = $this->getReadyRdKafkaConf($consumer['conf']);
+            $conf->setDefaultTopicConf($topicConf);
+
+            $kafkaConsumer = new \RdKafka\KafkaConsumer($conf);
+
+            $consumerDefinition->addMethodCall('setRdKafkaKafkaConsumer', [$kafkaConsumer]);
```
Comment (lnahiro) [inline DependencyInjection/M6WebKafkaExtension.php]: setConsumer ?

File: RdKafkaConsumerManager.php [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function setRdKafkaKafkaConsumer(\RdKafka\KafkaConsumer $kafkaConsumer)
+    {
+        $this->rdKafkaKafkaConsumer = $kafkaConsumer;
+    }
+
+    /**
+     * @param int $timeoutConsumingQueue
+     *
+     * @return $this
```
Comment (gmillet) [inline RdKafkaConsumerManager.php]: ?

File: RdKafkaConsumerManager.php [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+     * @return void
+     */
+    public function addTopic(array $name)
+    {
+        $this->rdKafkaKafkaConsumer->subscribe($name);
+    }
+
+    /**
+     * @param \RdKafka\KafkaConsumer $kafkaConsumer
+     * @return $this
```
Comment (gmillet) [inline RdKafkaConsumerManager.php]: ?

File: RdKafkaProducerManager.php [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+    /**
+     * @var int
+     */
+    protected $logLevel;
+
+    /**
+     * @var array
+     */
+    protected $brokers;
+    /**
```
Comment (gmillet) [inline RdKafkaProducerManager.php]: New line missing

File: RdKafkaProducerManager.php [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    /**
+     * @param string       $message
+     * @param integer      $partition
+     * @param integer|null $key
+     *
+     * @return callable
+     */
+    protected function produceForEachTopic(string $message, int $partition, string $key = null) : callable
```
Comment (gmillet) [inline RdKafkaProducerManager.php]: unused $key

File: Tests/Units/RdKafkaConsumerManager.php [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+    public function testShouldInformThatTheresATimeOut()
+    {
+        $this
+            ->given(
+                $eventDispatcherMock = $this->getEventDispatcherMock(),
+                $consumer = $this->getReadyBase($rdKafkaConsumerMock = $this->getRdKafkaConsumerMock(RD_KAFKA_RESP_ERR__TIMED_OUT), true, $eventDispatcherMock)
+            )
+            ->if($result = $consumer->consume())
+                ->then
+                ->string($result)
```
Comment (gmillet) [inline Tests/Units/RdKafkaConsumerManager.php]: Indentation

File: Tests/Units/RdKafkaConsumerManager.php [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+            ->given(
+                $eventDispatcherMock = $this->getEventDispatcherMock(),
+                $consumer = $this->getReadyBase($rdKafkaConsumerMock = $this->getRdKafkaConsumerMock('error'), true, $eventDispatcherMock)
+            )
+            ->exception(function () use ($consumer) {
+                $consumer->consume();
+            })
+                ->isInstanceOf('M6Web\Bundle\KafkaBundle\Exceptions\KafkaException')
+                ->hasMessage('error')
+                ->mock($eventDispatcherMock)
```
Comment (gmillet) [inline Tests/Units/RdKafkaConsumerManager.php]: Indentation

File: Tests/Units/RdKafkaConsumerManager.php [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+                    ->call('dispatch')
+                        ->never()
+                ->mock($rdKafkaConsumerMock)
```
Comment (gmillet) [inline Tests/Units/RdKafkaConsumerManager.php]: Indentation

File: Tests/Units/RdKafkaProducerManager.php [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+            ->given(
+                $producer = $this->getReadyBase(false),
+                $producer->setEventDispatcher($eventDispatcherMock)
+            )
+            ->exception(function () use ($producer) {
+                $producer->produce('message', 58);
+            })
+                ->isInstanceOf('M6Web\Bundle\KafkaBundle\Exceptions\KafkaException')
+                ->hasMessage('Random error from Kafka itself')
+                ->mock($eventDispatcherMock)
```
Comment (gmillet) [inline Tests/Units/RdKafkaProducerManager.php]: Indentation

File: RdKafkaConsumerManager.php [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     * @return \RdKafka\Message|string
+     * @throws KafkaException
+     */
+    public function consume()
+    {
+        $message =  $this->rdKafkaKafkaConsumer->consume($this->timeoutConsumingQueue);
+
+        if ($message->err === RD_KAFKA_RESP_ERR__PARTITION_EOF) {
+            return 'No more message';
```
Comment (lnahiro) [inline RdKafkaConsumerManager.php]: return null ?

File: DependencyInjection/M6WebKafkaExtension.php [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+        foreach ($config['consumers'] as $key => $consumer) {
+            $consumerDefinition = $this->getDefinition('M6Web\Bundle\KafkaBundle\RdKafkaConsumerManager');
+
+            $this->setEventDispatcher($config, $consumerDefinition);
+
+            $topicConf = $this->getReadyTopicConf($consumer['topicConf']);
+            $conf = $this->getReadyRdKafkaConf($consumer['conf']);
+            $conf->setDefaultTopicConf($topicConf);
+
+            $kafkaConsumer = new \RdKafka\KafkaConsumer($conf);
```
Comment (fabdsp) [inline DependencyInjection/M6WebKafkaExtension.php]: Can we configure this class name ?

File: RdKafkaConsumerManager.php [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        if ($message->err === RD_KAFKA_RESP_ERR__TIMED_OUT) {
+            return 'Time out';
```
Comment (lnahiro) [inline RdKafkaConsumerManager.php]: throw exception ?

File: RdKafkaProducerManager.php [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        return function ($topic) use ($message, $key, $partition) {
+            $topic->produce($partition, 0, $message);
```
Comment (lnahiro) [inline RdKafkaProducerManager.php]: comment to explain the second parameter

File: DependencyInjection/M6WebKafkaExtension.php [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+                return;
+            }
+
+            if ($error === RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS) {
+                $kafka->assign(null);
+
+                return;
+            }
+
+            throw new \Exception($error);
```
Comment (gmillet) [inline DependencyInjection/M6WebKafkaExtension.php]: Why not a less generic exception ?

File: IntegrationTests/Units/DependencyInjection/M6WebKafkaExtension.php [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $container = $this->getContainerForConfiguration('config');
+        $container->compile();
+
+        $this
+            ->boolean($container->has('m6_web_kafka.producer.producer1'))
+                ->isTrue()
+            ->object($producer = $container->get('m6_web_kafka.producer.producer1'))
+                ->isInstanceOf(RdKafkaProducerManager::class)
+            ->variable($producer->produce('\O/'))
```
Comment (gmillet) [inline IntegrationTests/Units/DependencyInjection/M6WebKafkaExtension.php:35]: Not tested

File: DependencyInjection/M6WebKafkaExtension.php [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+                ]
+            );
+        }
+    }
+
+    /**
+     * @param array      $entity
+     * @param Definition $entityDefinition
+     */
+    protected function setProducerManager(array $entity, Definition $entityDefinition)
```
Comment (fabdsp) [inline DependencyInjection/M6WebKafkaExtension.php]: configProducerManager ? or why not getProducerManager with the getDefinition inside and all the addMethodCall ?

File: IntegrationTests/Units/DependencyInjection/M6WebKafkaExtension.php [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+        $container = $this->getContainerForConfiguration('config');
+        $container->compile();
+
+        $this
+            ->boolean($container->has('m6_web_kafka.consumer.consumer1'))
+                ->isTrue()
+            ->object($consumer = $container->get('m6_web_kafka.consumer.consumer1'))
+                ->isInstanceOf(RdKafkaConsumerManager::class)
+            ->variable($message1 = $consumer->consume())
+            ->variable($message1->payload)
```
Comment (gmillet) [inline IntegrationTests/Units/DependencyInjection/M6WebKafkaExtension.php:39]: `->variable(($consumer->consume())->payload)`

File: README.md [hunk 32]
```diff
@@ -1,2 +1,170 @@
...
+    - batman
+    - catwoman
+ ```
+ 
+###Producer
+
+A producer will be used to send messages to the server.
+
+In the Kafka Model, messages are sent to topics partitioned and stored on brokers.
+It means that in the configuration for a producer you will have to specify the brokers, topics and optionnaly the log level and the strategy partitioner.
```
Comment (fabdsp) [inline README.md]: optionally

File: IntegrationTests/Units/DependencyInjection/M6WebKafkaExtension.php [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * @param $fixtureName
+     *
+     * @return ContainerBuilder
+     */
+    protected function getContainerForConfiguration(string $fixtureName): ContainerBuilder
+    {
+        $extension = new Base();
+        $parameterBag = new ParameterBag(array('kernel.debug' => true));
```
Comment (gmillet) [inline IntegrationTests/Units/DependencyInjection/M6WebKafkaExtension.php]: array() is so old 😄

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (gmillet) [reply DependencyInjection/Configuration.php]: `->defaultValue([])`

File: DependencyInjection/Configuration.php (hunk 3 above)
Comment (gmillet) [reply DependencyInjection/Configuration.php]: ->defaultValue([])

File: DependencyInjection/Configuration.php (hunk 5 above)
Comment (gmillet) [inline DependencyInjection/Configuration.php]: ->defaultValue([])

File: DependencyInjection/Configuration.php [hunk 34]
```diff
@@ -0,0 +1,78 @@
...
+                                ->prototype('scalar')->end()
+                            ->end()
+                            ->scalarNode('log_level')->defaultValue(LOG_WARNING)->end()
+                            ->arrayNode('topics')
+                                ->useAttributeAsKey('key')
+                                ->prototype('array')
+                                    ->children()
+                                        ->arrayNode('conf')
+                                            ->prototype('scalar')->end()
+                                            ->defaultValue(array())
```
Comment (gmillet) [inline DependencyInjection/Configuration.php]: ->defaultValue([])

File: IntegrationTests/Units/RdKafkaConsumerManager.php [hunk 35]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * @return void
+     */
+    public function testShouldConsumeMessages()
+    {
+        $topicConf = new \RdKafka\TopicConf();
+        $topicConf->set('auto.offset.reset', 'smallest');
+        $conf = new \RdKafka\Conf();
+        $conf->set('group.id', 'myConsumerGroup');
+
```
Comment (twdmnd) [inline IntegrationTests/Units/RdKafkaConsumerManager.php]: extra line

File: RdKafkaConsumerManager.php (hunk 24 above)
Comment (fabdsp) [inline RdKafkaConsumerManager.php]: Maybe null ? or we can set a reason somewhere ?

File: RdKafkaConsumerManager.php (hunk 26 above)
Comment (fabdsp) [inline RdKafkaConsumerManager.php]: null ? or we can set a reason somewhere ?

File: RdKafkaConsumerManager.php [hunk 36]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * @return \RdKafka\Message|string
+     * @throws KafkaException
+     */
+    public function consume()
+    {
+        $message =  $this->rdKafkaKafkaConsumer->consume($this->timeoutConsumingQueue);
+
+        if ($message->err === RD_KAFKA_RESP_ERR__PARTITION_EOF) {
```
Comment (gmillet) [inline RdKafkaConsumerManager.php]: switch ?

**Review by fabdsp — CHANGES_REQUESTED** [review]

File: RdKafkaConsumerManager.php (hunk 26 above)
Comment (NastasiaSaby) [reply RdKafkaConsumerManager.php]: @fabdsp  I have thought about sending back the null object message directly from the rdKafka, but finally I chose another option. Can we discuss about this point?

File: DependencyInjection/M6WebKafkaExtension.php (hunk 10 above)
Comment (NastasiaSaby) [reply DependencyInjection/M6WebKafkaExtension.php]: @fabdsp It was possible. Then, I thought it was weird. So I removed this option. Can we discuss about this point?

File: DependencyInjection/M6WebKafkaExtension.php (hunk 25 above)
Comment (NastasiaSaby) [inline DependencyInjection/M6WebKafkaExtension.php]: @fabdsp I don't understand :blush:

File: RdKafkaConsumerManager.php (hunk 26 above)
Comment (NastasiaSaby) [reply RdKafkaConsumerManager.php]: Yes, I will let the null RdKakfa\Message object

File: DependencyInjection/M6WebKafkaExtension.php (hunk 13 above)
Comment (NastasiaSaby) [reply DependencyInjection/M6WebKafkaExtension.php]: I find it ugly, but i guess it is not a good answer. So I changed it.

File: DependencyInjection/M6WebKafkaExtension.php [hunk 37]
```diff
@@ -172,9 +160,8 @@ protected function getReadyRdKafkaConf(array $confToSet): \RdKafka\Conf
...
         $rdKafkaConf = $this->getRdKafkaConf();
 
-        array_walk($confToSet, function ($confValue, $confIndex) use ($rdKafkaConf) {
-            $rdKafkaConf->set($confIndex, (string) $confValue);
-        });
+        $revertConfToSet = array_flip($confToSet);
+        array_walk($revertConfToSet, [$rdKafkaConf, 'set']);
 
         // Set a rebalance callback to log automatically assign partitions
         $rdKafkaConf->setRebalanceCb($this->handlePartitionsAssignment());
```
Comment (fabdsp) [inline DependencyInjection/M6WebKafkaExtension.php]: Don't you think it's a but strange to set a callable define in the extension class ? Maybe handlePartitionsAssignment could be static in its own class ?

File: RdKafkaConsumerManager.php [hunk 38]
```diff
@@ -70,21 +67,11 @@ public function consume()
...
-            return 'No more message';
-        }
-
-        if ($message->err === RD_KAFKA_RESP_ERR__TIMED_OUT) {
-            return 'Time out';
-        }
-
         if ($message->err === RD_KAFKA_RESP_ERR_NO_ERROR) {
-            $this->rdKafkaKafkaConsumer->commit();
+            $this->rdKafkaKafkaConsumer->commit($message);
```
Comment (fabdsp) [inline RdKafkaConsumerManager.php]: We want auto commit only ? What do you think of having a "autocommit" parameter for "consume" ? and a "commit" method ? Maybe I've missed something.

**Review by fabdsp — APPROVED** [review]
Sinon 👍

**Comment by fabdsp** [comment]
test this please

File: Tests/Units/RdKafkaConsumerManager.php [hunk 39]
```diff
@@ -53,8 +54,9 @@ public function testShouldInformThatTheresNoMoreMessage()
             )
             ->if($result = $consumer->consume())
             ->then
-                ->string($result)
-                    ->isEqualTo('No more message')
+                ->object($result)
```
Comment (gmillet) [inline Tests/Units/RdKafkaConsumerManager.php]: Useless ?

File: Tests/Units/RdKafkaConsumerManager.php [hunk 40]
```diff
@@ -75,9 +77,10 @@ public function testShouldInformThatTheresATimeOut()
                 $consumer = $this->getReadyBase($rdKafkaConsumerMock = $this->getRdKafkaConsumerMock(RD_KAFKA_RESP_ERR__TIMED_OUT), true, $eventDispatcherMock)
             )
             ->if($result = $consumer->consume())
-                ->then
-                ->string($result)
-                    ->isEqualTo('Time out')
+            ->then
+                ->object($result)
```
Comment (gmillet) [inline Tests/Units/RdKafkaConsumerManager.php]: Useless ?

File: Tests/Units/RdKafkaConsumerManager.php (hunk 39 above)
Comment (NastasiaSaby) [reply Tests/Units/RdKafkaConsumerManager.php]: Yes

File: Tests/Units/RdKafkaConsumerManager.php (hunk 40 above)
Comment (NastasiaSaby) [reply Tests/Units/RdKafkaConsumerManager.php]: Yes

File: .travis.yml [hunk 41]
```diff
@@ -7,11 +7,27 @@ branches:
...
+  # install librdkafka
+  - cd /tmp && git clone https://github.com/edenhill/librdkafka.git
+  - cd /tmp/librdkafka && ./configure
+  - make
+  - sudo make install
+
+  #install rdkafka
+  - cd /tmp && git clone https://github.com/arnaud-lb/php-rdkafka.git
+  - cd /tmp/php-rdkafka && phpize && ./configure
+  - sudo make all -j 5 && sudo make install
```
Comment (fabdsp) [inline .travis.yml]: make all sans le -j 5

**Review by fabdsp — APPROVED** [review]

File: .travis.yml [hunk 42]
```diff
@@ -7,11 +7,27 @@ branches:
...
+  - cd /tmp/librdkafka && ./configure
+  - make
+  - sudo make install
+
+  #install rdkafka
+  - cd /tmp && git clone https://github.com/arnaud-lb/php-rdkafka.git
+  - cd /tmp/php-rdkafka && phpize && ./configure
+  - sudo make all && sudo make install
+  - echo "extension = rdkafka.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
+  - cd /home/travis/build/M6Web/KafkaBundle
```
Comment (Oliboy50) [inline .travis.yml]: is this possible to use `TRAVIS_BUILD_DIR` instead of this path?

doc: https://docs.travis-ci.com/user/environment-variables/

**Review by Oliboy50 — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 42 above)
Comment (NastasiaSaby) [reply .travis.yml]: Of course

**Review by Oliboy50 — APPROVED** [review]

File: .travis.yml (hunk 42 above)
Comment (Oliboy50) [reply .travis.yml]: thanks :)

File: DependencyInjection/M6WebKafkaExtension.php [hunk 43]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        foreach ($config['producers'] as $key => $producer) {
+            $producerDefinition = $this->getDefinition('M6Web\Bundle\KafkaBundle\Manager\RdKafkaProducerManager');
+
+            $this->setEventDispatcher($config, $producerDefinition);
+
+            $conf = $this->getReadyRdKafkaConf($producer['conf']);
+            $rdKafkaProducer = new \RdKafka\Producer($conf);
+
+            $producerDefinition->addMethodCall('setRdKafkaProducer', [$rdKafkaProducer]);
```
Comment (lnahiro) [inline DependencyInjection/M6WebKafkaExtension.php]: setProducer ?

File: DependencyInjection/M6WebKafkaExtension.php [hunk 44]
```diff
@@ -0,0 +1,170 @@
...
+
+            $this->setEventDispatcher($config, $consumerManager);
+
+            $topicConf = $this->getReadyTopicConf($consumer['topicConf']);
+            $conf = $this->getReadyRdKafkaConf($consumer['conf']);
+            $conf->setDefaultTopicConf($topicConf);
+
+            $kafkaConsumer = new \RdKafka\KafkaConsumer($conf);
+
+            $consumerManager->addMethodCall('setRdKafkaKafkaConsumer', [$kafkaConsumer]);
```
Comment (lnahiro) [inline DependencyInjection/M6WebKafkaExtension.php]: setConsumer ?

File: Event/EventLog.php [hunk 45]
```diff
@@ -0,0 +1,35 @@
...
+
+use Symfony\Component\EventDispatcher\Event as SymfonyEvent;
+
+/**
+ * Class EventLog
+ * @package M6Web\Bundle\KafkaBundle\Event
+ *
+ * A class to send an event to say that messages were produced or consumed for monitoring
+ */
+class EventLog extends SymfonyEvent
```
Comment (lnahiro) [inline Event/EventLog.php]: KafkaEvent ?

File: Resources/config/services.yml [hunk 46]
```diff
@@ -0,0 +1,3 @@
+parameters:
+    m6_web_kafka.producer.class: 'M6Web\Bundle\KafkaBundle\ProducerManager'
```
Comment (lnahiro) [inline Resources/config/services.yml]: unused ?

**Pull request merged**